sw/host/opentitanlib/src/rescue/dfu.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::with_unknown;
+use anyhow::Result;
+use zerocopy::{AsBytes, FromBytes, FromZeroes};
+
+with_unknown! {
+    #[derive(FromBytes, FromZeroes, AsBytes)]
+    pub enum DfuState: u8 [default = Self::AppIdle] {
+      AppIdle = 0,
+      AppDetach= 1,
+      Idle = 2,
+      DnLoadSync = 3,
+      DnLoadBusy = 4,
+      DnLoadIdle = 5,
+      ManifestSync = 6,
+      Manifest = 7,
+      ManifestWaitReset = 8,
+      UpLoadIdle = 9,
+      Error = 10,
+    }
+}
+
+with_unknown! {
+    #[derive(FromBytes, FromZeroes, AsBytes)]
+    pub enum DfuError: u8 [default = Self::Ok] {
+      Ok = 0,
+      Target = 1,
+      File = 2,
+      Write = 3,
+      Erase = 4,
+      CheckErased = 5,
+      Prog = 6,
+      Verify = 7,
+      Address = 8,
+      NotDone = 9,
+      Firmware = 10,
+      Vendor = 11,
+      UsbReset = 12,
+      PowerOnReset = 13,
+      Unknown = 14,
+      StalledPkt = 15,
+    }
+}
+
+with_unknown! {
+    pub enum DfuRequest: u8 {
+      Detach = 0,
+      DnLoad = 1,
+      UpLoad = 2,
+      GetStatus = 3,
+      ClrStatus = 4,
+      GetState = 5,
+      Abort = 6,
+      BusReset = 7,
+    }
+}
+
+#[derive(Clone, Copy)]
+#[repr(u8)]
+pub enum DfuRequestType {
+    Out = 0x21, // direction=out, type=class, recipient=interface.
+    In = 0xA1,  // direction=in, type=class, recipient=interface.
+}
+
+impl From<DfuRequestType> for u8 {
+    fn from(val: DfuRequestType) -> Self {
+        val as u8
+    }
+}
+
+#[derive(Clone, FromBytes, FromZeroes, AsBytes, Default)]
+#[repr(C)]
+pub struct DfuStatus {
+    status: DfuError,
+    poll_timeout: [u8; 3],
+    state: DfuState,
+    string: u8,
+}
+
+impl std::error::Error for DfuError {}
+
+impl DfuStatus {
+    pub fn status(&self) -> std::result::Result<(), DfuError> {
+        match self.status {
+            DfuError::Ok => Ok(()),
+            e => Err(e),
+        }
+    }
+    pub fn poll_timeout(&self) -> u32 {
+        u32::from_le_bytes([
+            self.poll_timeout[0],
+            self.poll_timeout[1],
+            self.poll_timeout[2],
+            0,
+        ])
+    }
+    pub fn state(&self) -> DfuState {
+        self.state
+    }
+    pub fn string(&self) -> u8 {
+        self.string
+    }
+}
+
+pub trait DfuOperations {
+    fn write_control(
+        &self,
+        request_type: u8,
```

Comment:
What is the purpose of maintaining the `request_type` parameter in `write_control` and `read_control`? 

Are there cases where we may want to set `read_control` with `DfuRequestType::Out` for example?

Reply:
The `request_type` field is part of the USB standard type identification for SETUP_DATA.  In addition to encoding direction, the request_type also encodes the protocol layer/destination of the request.  e.g.: Is this SETUP a standard/class/vendor request and is the destination device/interface/endpoint/other?

There is one case where we do want to use values other than `DfuRequestType::{In, Out}`: in the SPI implementation, there is a proprietary command to perform `Rescue::set_mode`.  This is needed because in the SPI implementation, it doesn't make sense to implement a bunch of the standard USB framework.